tests/test_mem_eff_attention.py

```diff
@@ -3,34 +3,120 @@
 # This source code is licensed under the BSD license found in the
 # LICENSE file in the root directory of this source tree.
 
+import math
+
 import pytest
 import torch
+from scipy.stats import binom_test
 
 import xformers.ops
 
+torch.backends.cuda.matmul.allow_tf32 = False
+cuda_only = pytest.mark.skipif(not torch.cuda.is_available(), reason="requires CUDA")
 _devices = ["cpu", "cuda"] if torch.cuda.is_available() else ["cpu"]
 
 
-def ref_attention(q, k, v):
+def assert_allclose(
+    out: torch.Tensor,
+    ref: torch.Tensor,
+    msg: str = "failed",
+    atol: float = 1e-8,
+    rtol: float = 1e-5,
+) -> None:
+    flatten_diff = ((out - ref).abs() - atol - ref.abs() * rtol).flatten()
+    max_pos = flatten_diff.argmax()
+    assert torch.allclose(out, ref, rtol=rtol, atol=atol), (
+        f"{msg}: max_diff={flatten_diff.max()}: "
+        f"out={out.flatten()[max_pos]} and ref={ref.flatten()[max_pos]} "
+        f"/ atol={atol}, rtol={rtol}"
+    )
+
+
+def ref_attention(q, k, v, attn_bias=None, drop_mask=None, p=0.0):
+    q = q.float()
+    k = k.float()
+    v = v.float()
+
     q = q * (1 / q.shape[-1] ** 0.5)
-    return (q @ k.transpose(-2, -1)).softmax(-1) @ v
+    attn = q @ k.transpose(-2, -1)
+    if attn_bias is not None:
+        if isinstance(attn_bias, xformers.ops.AttentionMask):
+            attn_bias = attn_bias.to_tensor()
+        attn = attn + attn_bias.float()
+    attn = attn.softmax(-1)
+    if drop_mask is not None:
+        attn = attn * (drop_mask / (1 - p))
+    return attn @ v
 
 
-@pytest.mark.parametrize("k_len", [5, 6, 32])
+def create_attn_bias(
+    bias_type, batch_size: int, q_len: int, kv_len: int, device, dtype
+):
+    if bias_type is None:
+        return None
+    if bias_type is torch.Tensor:
+        attn_bias = torch.randn((batch_size, 1, kv_len), device=device, dtype=dtype) * 3
+        return attn_bias.expand(batch_size, q_len, kv_len)
+    if bias_type is xformers.ops.LowerTriangularMask:
+        return bias_type([batch_size, q_len, kv_len], dtype=dtype, device=device)
+    assert False, f"Unsupported bias type: {bias_type}"
+
+
+@pytest.mark.parametrize(
+    "attn_bias_type", [None, xformers.ops.LowerTriangularMask, torch.Tensor]
+)
+@pytest.mark.parametrize("k_len", [5, 6, 32, 128])
```

Comment:
nice test cascade ! that's some serious coverage

Reply:
Yeah, there are a lot of combinations being tested now. Tests are not instantaneous now (~1min ?), but it's not too bad I think